### DISCUSSION TITLE
Open MPI OSC pt2pt component and MPI_THREAD_MULTIPLE

### DISCUSSION THREAD

**Discussion by piyueh** [mpi4py/mpi4py#139] [Q&A] [answered]
Hi, I encountered this error when running a one-side communication code with CuPy:
```
The OSC pt2pt component does not support MPI_THREAD_MULTIPLE in this release.
Workarounds are to run on a single node, or to use a system with an RDMA
capable network such as Infiniband.
```
I'm not sure if this is due to I didn't write the code correctly (though the code worked correctly when using NumPy). This is a simple test code to reproduce the error:
```python
import cupy
from mpi4py import MPI
from mpi4py.util.dtlib import from_numpy_dtype

assert MPI.COMM_WORLD.size == 2, "This test requires exactly 2 ranks."
comm = MPI.COMM_WORLD
cupy.cuda.runtime.setDevice(comm.rank)

data = cupy.zeros((7, 7), dtype=int)
data[2:-2, 2:-2] = cupy.arange(9).reshape(3, 3) + 10 * comm.rank
cupy.cuda.get_current_stream().synchronize()

win = MPI.Win.Create(data, comm=comm)
win.Fence()
if comm.rank == 0:
    stype = from_numpy_dtype(data.dtype).Create_subarray((7, 7), (3, 2), (2, 3)).Commit()
    rtype = from_numpy_dtype(data.dtype).Create_subarray((7, 7), (3, 2), (2, 0)).Commit()
    win.Put([data, 1, stype], 1, [0, 1, rtype])
    win.Fence()
else:
    win.Fence()
    print(data)
win.Free()
```

The test was launched with
```shell
$ mpiexec -n 2 --mca opal_cuda_support 1 python test.py
```

In addition to the error message mentioned at the beginning of this post, a `mpi4py.MPI.Exception: MPI_ERR_WIN: invalid window` error was also raised at the line of `MPI.Win.Create`.

I'm using everything from Anaconda:
- `mpi=1.0=openmpi`   conda-forge
- `mpi4py=3.1.2=py39hc649577_0` from conda-forge
- `openmpi=4.1.1=hbfc84c5_0` from conda-forge
- `cupy=9.5.0=py39h499daff_1` from conda-forge
- `cudatoolkit=11.2.72=h2bc3f7f_0` from nvidia

The system is a single node with multiple K40 GPUs.
Thanks!

**Comment by piyueh** [comment]
Ok, I now can make the script run and get the correct answer by setting `mpi4py.rc.threads = False` or by the environment variable `MPI4PY_RC_THREADS=0`. However, I'm not sure I understand why we need multiple threads by default, and I don't know if disabling multithread will cause any performance penalty.

I guess this is more like a OpenMPI issue?

**Comment by leofang** [comment]
@piyueh Thank you for correctly raising a GH discussion for this topic. It is indeed Open MPI's implementation issue (and was a long-standing one), see the struggle they had that eventually led to the decision of not supporting threading for osc pt2pt: https://github.com/open-mpi/ompi/issues/2614#issuecomment-376289727. How about change `--mca osc` to something else (say `ucx`)?

**Reply by leofang** [reply]
See also the discussion in #34.

**Comment by dalcinl** [comment, marked as answer]
mpi4py initializes the MPI runtime automatically. As I cannot possibly know how users will use MPI, I decided to request multiple thread support by default, simply because that `MPI_THREAD_MULTIPLE` mode is the one that will guarantee the MPI runtime working properly on all circumstances. You may not need MPI+threads, but others may, so better to be safe than sorry. 

Years ago I asked Bill Gropp (former MPICH lead developer) for advice about this subject, and he basically agreed that using `MPI_THREAD_MULTIPLE` was the right thing to. Bill also said that the performance overhead for single-threaded applications should be minimal. Years after that exchange with Bill, I added the `mpi4py.futures` module, which uses threading and calls into MPI from multiple threads.

PS: IIRC, Open MPI v5 will remove the osc p2pt component.

**Comment by piyueh** [comment]
@leofang @dalcinl Thanks for the reply! I tried both `ucx` and `rdma` for the `osc` framework. When threading support is enabled, both `ucx` and `rdma` didn't work with the CuPy code I provided. However, when running CPU code (i.e., replacing `cupy` with `numpy`) `ucx` worked in the situation mentioned in #34 (i.e., no explicit mpi wrapper), and both worked when using `mpiexec -n 1` (`pt2pt` didn't even work in this case).

I admit I didn't spend much time looking into properly using `ucx` and `rdma`. I might miss some parameters or something else to properly use `ucx` or `rdma` with CUDA.

For others' information in the future, this may be relevant: https://github.com/openucx/ucx/issues/7538
